libkbfs/delayed_cancellation.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+)
+
+const (
+	// CtxReplayKey is a context key for CtxReplayFunc
+	CtxReplayKey = "libkbfs-replay"
+
+	// CtxCancellationDelayerKey is a context key for using cancellationDelayer
+	CtxCancellationDelayerKey = "libkbfs-cancellation-delayer"
+)
+
+// CtxReplayFunc is a function for replaying a series of changes done on a
+// context.
+type CtxReplayFunc func(ctx context.Context) context.Context
+
+// CtxNotReplayableError is returned when NewContextWithReplayFrom is called on
+// a ctx with no replay func.
+type CtxNotReplayableError struct{}
+
+func (e CtxNotReplayableError) Error() string {
+	return "Unable to replay on ctx"
+}
+
+// NoCancellationDelayerError is returned when EnableDelayedCancellationWithGracePeriod or
+// ExitCritical are called on a ctx without Critical Awareness
+type NoCancellationDelayerError struct{}
+
+func (e NoCancellationDelayerError) Error() string {
+	return "Context doesn't have critical awareness or CtxCancellationDelayerKey " +
+		"already exists in ctx but is not of type *cancellationDelayer"
+}
+
+// ContextAlreadyHasCancellationDelayerError is returned when
+// NewContextWithCancellationDelayer is called second time on the same ctx, which
+// is not supported yet.
+type ContextAlreadyHasCancellationDelayerError struct{}
+
+func (e ContextAlreadyHasCancellationDelayerError) Error() string {
+	return "Context already has critical awareness; only one layer is supported."
+}
+
+// NewContextReplayable creates a new context from ctx, with change applied. It
+// also makes this change replayable by NewContextWithReplayFrom. When
+// replayed, the resulting context is replayable as well.
+//
+// It is important that all WithValue-isu mutations on ctx is done "replayably"
+// (with NewContextReplayable) if any delayed cancellation is used, e.g.
+// through EnableDelayedCancellationWithGracePeriod,
+func NewContextReplayable(
+	ctx context.Context, change CtxReplayFunc) context.Context {
+	var replay CtxReplayFunc
+	replay = func(ctx context.Context) context.Context {
+		ctx = change(ctx)
+		// _ is necessary to avoid panic if nil
+		replays, _ := ctx.Value(CtxReplayKey).([]CtxReplayFunc)
+		replays = append(replays, replay)
```

Comment:
I couldn't thinking of a reason why not so I implemented your idea. Thanks Jeremy!